### ISSUE TITLE
Asynchronous Driver For Python <Tornado and Tulip>

### ISSUE THREAD

**Issue by v3ss0n** [rethinkdb/rethinkdb#2622]
Change feed on 1.13 is very intersting feature.Multi-media-realtime HTML5 projects are getitng very popular these days , and i am building one right now too , a multi-media-webchat .It needs a lot of serverside notifications.
However , RethinkDB still needs async driver for Tornadoweb, like Motor driver by MongoDB does , that is only deal-blocking  feature that blocking me changing from Mongo.
I Love ReQL and I really believe that RethinkDB is NoSQL-Done-Right. 

Any Pointers on modifying current driver work async would be very good,I would like to contribute. (I am still quite new on Async and Tornado.

**Comment by kzahel** [comment]
Hi I'm just evaluating rethinkDB at this point for use in a new project. Having support for an asynchronous tornado driver would be very compelling for me. Out of the box support for a python api that mirrors the node/javascript API would be really great.

**Comment by danielmewes** [comment]
We're going to consider different options (including Tornado and Python 3 asyncio) for making changefeeds work better in Python as part of #3298. This will happen very soon.
I'm not sure yet to what degree we are going to convert the rest of the driver (apart of changefeeds) to an asynchronous interface.

**Comment by techdragon** [comment]
Ideally the outcome would be to refactor the python code to be event loop agnostic. The same "protocol" code reused in event loop drivers so their feature compatibility is ensured. I feel I have to point this out because unfortunately few libraries are written this way. I dont want to see a future where I'll be installing with ```pip install rethinkdb-twisted``` in one codebase and ```pip install rethinkdb-asyncio``` in another.

The Autobahn project has a great structure for this kind of thing,  https://github.com/tavendo/AutobahnPython/

The Autobahn code is structured so if/when people need to use the protocol with some other unsupported event loop, they just have to rewrite the much smaller driver code to use their event loop.
In the case of rethinkdb there would be an additional 'synchronous' driver branch.

**Comment by csytan** [comment]
I think the cleanest way to get the driver to run on both Python 2 & 3 would be to use Tornado. Tornado coroutines are very similar to Asyncio's and are also compatible with [Asyncio's event loop](http://tornado.readthedocs.org/en/latest/asyncio.html).

**Comment by danielmewes** [comment]
@techdragon We'd like to avoid duplication as well. Will see what we can do.
@csytan That's interesting, thanks for pointing it out.

Moving this to 2.1. We'll try to get Tornado integration done soon.

**Comment by deontologician** [comment]
Just a note from my personal experience, Tornado is the node equivalent right now in the python world. Twisted has been around longer, but isn't as widely used in the web-app community. In terms of which async loop to support first, Tornado would be a solid choice.

**Comment by danielmewes** [comment]
@deontologician Ok that's good to know. I have no personal preference with respect to whether to support Twisted or Tornado first. My understanding from what @larkost said is that Twisted tends to be slightly easier to install due to more widely available packages.

If Tornado is easier to support than Twisted that sounds like we should support Tornado first. Otherwise we should just pick one and then follow up the other one soon (maybe we can even do both for 2.0).
@gchpaco any opinions?

**Comment by deontologician** [comment]
I can't comment on which one is easier to implement, just in terms of who is most likely to make use of it

**Comment by gchpaco** [comment]
It has again been a while since I did bleeding edge Python, but if @csytan is right then it would be *much* less work to support Tornado-and-asyncio-etc than it would be to support only Twisted.

@techdragon That's interesting!  I'm going to have to look at it in more detail.  Certainly a possibility.

**Comment by danielmewes** [comment]
Ok let's to Tornado first then.

**Comment by techdragon** [comment]
@gchpaco & @danielmewes - If you refactor the logic to separate the drivers from the core logic the drivers rely on, then at least in theory, each driver should be simpler to build. Tornado is the current favourite among a lot of developers, and would be great, but when you refactor the logic you need to take into account that you need to maintain the existing synchronous driver. 

Twisted has a LOT of documentation regarding its use.

I would try to refactor the current code to be broken apart like Autobahn has theirs, supporting twisted and regular synchronous python first, then adding tornado & asyncio in whichever order works best. 
I dont think the drivers will be the hard part of getting this done, I would expect that after the logic refactoring, the drivers, "Synchronous", "Twisted", "Asyncio", and "Tornado" will all be fairly strait forward.

**Comment by v3ss0n** [comment]
Thanks a lot for this going forward.I haven't check for long due to being busy with Async chat Implementation on Tornado and Mongo.

@danielmewes  , what @csytan  true that tornado's coroutines are compatible with asyncio of Python3. It gives you clean code using yield, without needing to worry about callbacks , and thats the reason for web devs choosing tornado over twisted.
Tornado have a twisted bridge , I think its bidrectional so if you support tornado , it will support twisted too.

One Stone , a flock of birds. Nice? :)

**Comment by v3ss0n** [comment]
@techdragon  what do you think about tornado twisted bridge? With that , supporting tornado will help twisted community atonce too. I have't tested it personally.

**Comment by ajdavis** [comment]
Hi, I'm a PyMongo committer and Motor's author. Motor supports Tornado using greenlets (so it falls somewhere in the middle of the spectrum between an async rewrite of PyMongo vs. running PyMongo in a thread pool), and this year it will support asyncio as well. I plan to abstract away the differences between asyncio and Tornado, and eventually do the same for Twisted. You can see the work in progress here:

https://github.com/mongodb/motor/tree/asyncio/motor/frameworks

If you have any questions I'm happy to offer **very small** amounts of time to answer them. =)

**Comment by danielmewes** [comment]
Thanks @ajdavis for chiming in. We'll have a look at how you're solving it in Motor. :)

**Comment by v3ss0n** [comment]
Thanks alot @ajdavis !i am fan of your works. Both motor and toro!

**Comment by v3ss0n** [comment]
Hello guys , if this work is started at python driver , notify me , i would like to take a look too.
Thanks!

**Comment by danielmewes** [comment]
Thank you @v3ss0n ! @gchpaco was going to look into it soon.

**Comment by danielmewes** [comment]
Similar to https://github.com/rethinkdb/rethinkdb/issues/3678#issuecomment-75142374 , we might have to pass the callback into `run` itself, rather than having an `on` method on the cursor. This is to avoid the problem that by default `run` doesn't return anything at all until it has gotten the first batch, which would be a problem here.

**Comment by coffeemug** [comment]
That sounds like a really good idea to me, for two reasons:

- It reduces the API footprint (there is no special `on` command to learn)
- It makes the API more consistent with the node drivers

**Comment by danielmewes** [comment]
From reading http://www.tornadoweb.org/en/stable/guide/async.html it sounds like the recommended interface for async APIs are futures.

So I imagine it would look something like this:
```
# returns a tornado.concurrent.Future on the result of run
run_ft = r.table(...).changes().run(conn, future=True)

# throws an exception on error, otherwise returns the run result (in this case a cursor)
feed = run_ft.result()

while True:
  next_ft = feed.next(future=True)
  print next_ft.result()
```

You can also specify a timeout when calling `result()` on the future. See http://www.tornadoweb.org/en/stable/concurrent.html#tornado.concurrent.Future for details.

There also is a `Future.add_done_callback()` method that can be used if you prefer using a callback.

**Comment by danielmewes** [comment]
@v3ss0n since you've actually worked with Tornado, does that sounds reasonable to you (using Futures like described above)?

**Comment by ajdavis** [comment]
Let me correct this a little bit. Yes, "run" should return a Future, but to wait for the future to resolve you must yield it in a coroutine:
```
from tornado import gen

@gen.coroutine
def f():
    feed = yield r.table(...).changes().run(conn, future=True)
```
Now, if you want to iterate over results that can be tricky in a coroutine, since there's no place for a "yield" within a Python "for" loop. In Motor I came up with this API:

http://motor.readthedocs.org/en/stable/api/motor_cursor.html#motor.MotorCursor.fetch_next

**Comment by danielmewes** [comment]
Thanks so much for the info @ajdavis.
The Motor API for solving the loop problem looks nice.

So to update the complete example with your comments, we would have
```py
#... imports ...

@gen.coroutine
def f():
  # throws an exception on error, otherwise returns the run result (in this case a cursor)
  feed = yield r.table(...).changes().run(conn, future=True)

  while (yield feed.fetch_next):
    # Is guaranteed to have a result since we have already waited on feed.fetch_next
    print feed.next()

IOLoop.current().run_sync(f)
```

Does that look right?

(sorry for my silly mistakes and questions, I'm pretty unexperienced with Python)

**Comment by ajdavis** [comment]
Yes, exactly right.

**Comment by danielmewes** [comment]
Let's start implementing support for this with an API like in this example https://github.com/rethinkdb/rethinkdb/issues/2622#issuecomment-75351376 .

We can still change some details about the API later if we want to (function names etc.), but it's important that we get the inner workings in place.

**Comment by danielmewes** [comment]
(@gchpaco has already started looking into this)

**Comment by coffeemug** [comment]
Why not do this:

```py
#... imports ...

@gen.coroutine
def f():
  # throws an exception on error, otherwise returns the run result (in this case a cursor)
  feed = yield r.table(...).changes().run(conn, future=True)

  for (x in yield feed.next()):
    # Is guaranteed to have a result since we have already waited on feed.fetch_next
    print x

IOLoop.current().run_sync(f)
```

We already told run that it should return a future, so `fetch_next` seems unnecessary. We can just change the implementation of `next` in this case.

**Comment by coffeemug** [comment]
Err, sorry, the whole `in` thing wouldn't work here (I haven't touched Python in a while). But there must be another pattern here. My point was that we should avoid `fetch_next` if possible.

**Comment by danielmewes** [comment]
@coffeemug `fetch_next` duplicates what @ajdavis did in the Motor API, see https://github.com/rethinkdb/rethinkdb/issues/2622#issuecomment-75350724 .

If we can find a nicer way around that Python limitation that would be good. Not sure if it exists.
(@larkost @timmaxw @Tryneus @gchpaco @v3ss0n any other suggestions?)

**Comment by coffeemug** [comment]
Actually wait, the `in` thing might work.

**Comment by v3ss0n** [comment]
Sorry i was a bit busy.

you can iterate the yield result.

Here is how we do with tornado and Motor:
```python
@gen.coroutine
def get():
    cursor = db.test_collection.find({'i': {'$lt': 5}})
    for document in (yield cursor.to_list(length=100)):
        print document
```

**Comment by danielmewes** [comment]
Thanks @v3ss0n. Using `to_list` I presume loads all data into the client at once.

This isn't quite the same as iterating over the cursor directly which would fetch results in smaller batches only when they are needed. For example you can iterate over a data set that doesn't even fit into memory on the client with a cursor, but not with `to_list`.

We still need a way for iterating over a cursor directly.

@coffeemug Have you found out whether the `in` actually works with `yield` on individual items?

**Comment by coffeemug** [comment]
No, I didn't. Someone who knows Python a little better should look at the API. But you can definitely do this with iterators and generators combo. For example `for x in xrange(10): ...` works, so in our case `for x in feed:` should work too. The difference is that the feed would yield internally (as long as you set `future=True`).

**Comment by danielmewes** [comment]
http://tornado.readthedocs.org/en/latest/guide/coroutines.html#looping supports what @ajdavis said that you cannot use `yield` in a loop like that. Assuming that documentation is correct.
Unfortunately I don't know enough about Python and Tornado to understand why this is the case. (@gchpaco, @v3ss0n, @ajdavis any insights?)

**Comment by v3ss0n** [comment]
The problem with python is yielding each and every iterated result. So we can't do that.
In motor , work around is to yield ```cursor.fetch_next``` and get value using ```.next_object()```
```python
 @gen.coroutine
 def do_find():
     cursor = db.test_collection.find({'i': {'$lt': 5}})
     while (yield cursor.fetch_next):
         document = cursor.next_object()
         print document
```

**Comment by ajdavis** [comment]
Right, this yields once:
```
for document in (yield cursor.to_list(length=100)):
        print document
```
By the time the yield statement is complete, Motor has loaded 100 documents in memory. Then you just have a list and iterate it, without yielding again. So to stream an unlimited number of documents with bounded memory, you need another loop outside *that*, doing something like "while cursor.alive" (I haven't tested that recently).

So why does yield seem to act differently in a while loop than in a for loop? Why can one yield statement be executed N times in a while loop, but only once in a for loop?
```
>>> def whileloop():
...     while (yield future):
...         pass
...
>>> def forloop():
...     for i in (yield future):
...         pass
...
>>> import dis
>>> dis.dis(whileloop)
  2           0 SETUP_LOOP              11 (to 14)
        >>    3 LOAD_GLOBAL              0 (future)
              6 YIELD_VALUE
              7 POP_JUMP_IF_FALSE       13

  3          10 JUMP_ABSOLUTE            3
        >>   13 POP_BLOCK
        >>   14 LOAD_CONST               0 (None)
             17 RETURN_VALUE
>>> dis.dis(forloop)
  2           0 SETUP_LOOP              15 (to 18)
              3 LOAD_GLOBAL              0 (future)
              6 YIELD_VALUE
              7 GET_ITER
        >>    8 FOR_ITER                 6 (to 17)
             11 STORE_FAST               0 (i)

  3          14 JUMP_ABSOLUTE            8
        >>   17 POP_BLOCK
        >>   18 LOAD_CONST               0 (None)
             21 RETURN_VALUE
```
At the end of the while loop we jump up to instruction pointer 3, which is before the YIELD_VALUE instruction. But at the end of the for loop we jump to instruction 8, which is after YIELD_VALUE.

This makes sense. After all, you expect you can stop a while loop by updating a global:
```
stop = False

def whileloop():
    while not stop:
        pass

def target():
    global stop
    time.sleep(1)
    stop = True

threading.Thread(target=target).start()
whileloop()
````
So the while loop reloads the value of its control variable on each iteration. But a for loop certainly can't reload its iterable and call iter() on it on each iteration, or it would never get past the first element.

**Comment by danielmewes** [comment]
Thank you so much @ajdavis for the in-depth explanation.

What if instead of a generator, we implement a special iterator for cursors?

```py
@gen.coroutine
def f():
  feed = yield r.table(...).changes().run(conn, future=True)

  for x in feed:
    print x
```

where `feed.__iter__()` would create something like this:
```py
class cursor_iterator(object):
  def __init__(self, feed):
    self.feed = feed

  def next(self):
    return yield self.feed.next()
```

Is this possible? I might be misunderstanding how these things are working.

**Comment by v3ss0n** [comment]
Not sure i got what you want to achieve but in Tornado's gen.coroutine to be asynchronus  , you will need to return a Future().
here is my channel object code from my chat server. This waits for multiple futures and notify specified group of futures , acting as channels.

``` python
class Channel(object):

    def __init__(self):
        self.waiters = defaultdict(deque) 

    def listen(self, uid=None):
        result = Future()
        self.waiters[uid].append(result)
        logging.info("waiting at topic %s" % uid)
        return result

    def notify(self, uid=None, msg=None):
        waiters = self.waiters[uid]
        for future in waiters:
            future.set_result(msg)
            self.waiters[uid] = deque()

class ChannelList(TornadoMsgHandler):
    @gen.coroutine
    def post(self, topic):
               ...... 
                msg = yield waiter.listen(topic)

```

**Comment by ajdavis** [comment]
No coroutine can do I/O without executing a yield statement. Look at your code: might it need to talk to the server? Then it must yield. A statement like "for x in feed" might need to do I/O to fetch more things, so it must contain a "yield" statement.

So must its caller, and so on up the chain, so that ultimately you have yielded to the event loop. Otherwise the loop cannot make progress and the I/O does not complete.

I was like you, once: I thought I could find the way. High and low I sought. But at the end of my seeking, I chose not merely to accept this commandment as a law of nature, but indeed to embrace it was the way of salvation.

https://glyph.twistedmatrix.com/2014/02/unyielding.html

**Comment by danielmewes** [comment]
@ajdavis 
> So must its caller, and so on up the chain, so that ultimately you have yielded to the event loop. Otherwise the loop cannot make progress and the I/O does not complete.

Ah, that was the thing I was missing. Thanks again @v3ss0n and @ajdavis (and sorry you guys need to explain Python basics to me)!

**Comment by v3ss0n** [comment]
@danielmewes no problem , I am happy to help.
Also check at tornado mailing list , we have python asynchronous  experts like @ajdavis and @bdernell , owner of Tornado active there. https://groups.google.com/forum/#!forum/python-tornado
i am gonna sleep for a while. i be back in 10 hrs.

**Comment by v3ss0n** [comment]
I am watching @gchpaco  's commit and seems good.
If its ready for testing?

**Comment by gchpaco** [comment]
@v3ss0n Not yet.  I hope to get it into a position to be banged on by people later tonight.

**Comment by gchpaco** [comment]
Okay, CR 2714.

**Comment by coffeemug** [comment]
:clap:

**Comment by v3ss0n** [comment]
Awesome~!  Going to test it soon.

**Comment by v3ss0n** [comment]
@gchpaco  thats mean merged to Rehtinkdb? i don't see anything new in python except repl.

**Comment by larkost** [comment]
@v3ss0n: that means that it is up for internal review. Once it makes it thought that, then it will be merged into the main branch, or rather branches in this case.

**Comment by v3ss0n** [comment]
@larkost  ah i c now

looking at :
https://github.com/rethinkdb/rethinkdb/commits/gchpaco_2622/drivers/python

**Comment by coffeemug** [comment]
@gchpaco -- a couple of questions:

- Could you post some minimal sample code here of how a user would use the driver with the async API?
- Would the old synchronous code still work without modifications?
- Since we have time before the release, could you start working on adding other Python async backends (starting with most common) when the code review goes through?

**Comment by coffeemug** [comment]
Specifically, with sample code it would be great to mirror @mlucy's mini-tutorial here https://github.com/rethinkdb/rethinkdb/issues/3678#issuecomment-78790839 in Python, so we understand how to use the API in different circumstances, how error handling works, etc.

**Comment by gchpaco** [comment]
Right now I'm trying to get it through review, but I'll try to do that.  As a temporary stopgap, see test/rql_test/connections/tornado_connection.py which is an analog of connection.py.

**Comment by v3ss0n** [comment]
https://github.com/rethinkdb/rethinkdb/blob/68551064ddfb61a77735875adcc070cf77798eb4/test/rql_test/connections/tornado_connection.py

this one right?

**Comment by gchpaco** [comment]
Yup.  Note that @Tryneus is working on refactoring some of it, so there may be some large changes when that goes through.

**Comment by Tryneus** [comment]
The refactor @gchpaco mentioned is up in review 2732, which is based of his branch and eclipses the previous review.  It is available in the branch `grey_issue_2622`.

**Comment by Tryneus** [comment]
Ok, the python driver refactor has been approved and merged to `next` in commits 73d83e8 (@gchpaco's changes), 154efa8 (driver refactor), and 43cd9d7 (test updates following the refactor), and `v2.0.x` in commits a126ed8, 7e51398, and a3d41d0.  Will be in release 2.0.

**Comment by v3ss0n** [comment]
Thank you so much , i like the way RethinkDB team handle projects , all official workflow integrated with github, that's so cool. 
I learn a lot from you and will practice in my company.
I am pulling now and going to test. If everything ok i will make a rehtinkdb + tornado example chat demo.

**Comment by danielmewes** [comment]
Out of curiosity @gchpaco:
How does this work exactly (example from https://github.com/rethinkdb/docs/issues/683)?
```rb
    # Print every row in the table.
    for future in (yield r.table('test').order_by(index='id').run(connection)):
        item = yield future
        print(item)
```

More specifically: how do we know in advance whether the cursor will have another result in the for loop? Does `yield future` always wait for the next batch to be loaded unless the cursor is a changefeed?

**Comment by Tryneus** [comment]
@danielmewes, I just tested this and it isn't the expected behavior (this is my fault due to some of the cursor changes in the refactor).  `yield future` will raise a `StopIteration` exception once it is past the end of the cursor.  Unfortunately, if the user doesn't catch this, it interacts poorly with Tornado, which assumes the `StopIteration` means that the coroutine has finished.

Opened #3974 for this issue.

**Comment by Tryneus** [comment]
So based on my suggestion in #3974, that loop would throw `RqlCursorEmpty` after the end of the cursor.

**Comment by danielmewes** [comment]
Having to handle the `RqlCursorEmpty` exception for every loop over a cursor sounds a little annoying.

Should we add a `fetch_next` like in Motor to guarantee that the next call to `next` doesn't throw?

```py
cursor = yield r.table('test').order_by(index='id').run(connection)
while (yield cursor.fetch_next):
    item = yield cursor.next()
    print(item)
```

**Comment by Tryneus** [comment]
@danielmewes, I think that would be useful, I'll do it alongside #3974.

**Comment by danielmewes** [comment]
:+1:

**Comment by ajdavis** [comment]
Note in Motor, you can *create* the cursor without a yield since it hasn't begun I/O yet, just stored the query in a MotorCursor object. Only on the first `yield cursor.fetch_next` do we send the query to the server and get the first batch:

http://motor.readthedocs.org/en/stable/api/motor_cursor.html#motor.MotorCursor.fetch_next

Also, I chose the method name `next_object` instead of `next` to avoid interacting with Python's standard iterator protocol in any unexpected way, since a MotorCursor *cannot* act like a standard iterator:

https://docs.python.org/2/library/stdtypes.html#iterator-types

**Comment by danielmewes** [comment]
Thanks @ajdavis .
For RethinkDB we need to yield on creating the cursor because we don't know whether a given query actually returns one before getting the first response from the server (some queries return a single value, which we handle differently).

Good point about the incompatibility with the iterator interface.
We could certainly use a special `next` function for Tornado cursors (not sure how we would call it). I can't tell whether it's important enough to warrant the separate function though.
I wonder how exactly a script fails if it tries to use a Tornado cursor as an iterator. I'll try that later.

**Comment by Tryneus** [comment]
@danielmewes, it's would be fine to iterate over a `TornadoCursor`, as long as you didn't do it eagerly (you would end up with an infinite loop that would probably be OOM-killed), and as far as I can tell, we can't safely do a comprehension like `arr = [(yield x) for x in cursor if (yield cursor.fetch_next())]`.  Taking this into account, we should probably remove iteration from the `TornadoCursor` implementation.

**Comment by v3ss0n** [comment]
No problem now , this code works well. 

```python
con = yield conn
curs = yield evt.run(con)
messages = []
while (yield curs.fetch_next()):
    item = yield curs.next()
    messages.append(item)
```

**Comment by danielmewes** [comment]
@v3ss0n I can't see anything obviously wrong about your insert code.

Which revision of the RethinkDB repository are you on (`$ git status`)? I wonder if you might be affected by https://github.com/rethinkdb/rethinkdb/issues/3998. It's closed in the latest revision which you should get through `git pull`.

**Comment by v3ss0n** [comment]
i am at 01fa526b696fe0395fb1790e56c0c6b94791fb4d

**Comment by v3ss0n** [comment]
what branch i should pool? next or v2.0x ?

**Comment by danielmewes** [comment]
@v3ss0n Both branches should be ok. The commit you mentioned actually includes the fix for that issue. So something else must be going on. I opened https://github.com/rethinkdb/rethinkdb/issues/4007 for further investigation into this.

**Comment by shivekkhurana** [comment]
After reviewing a lot of amazing work by David Beazley (https://github.com/dabeaz), and with the improvements in async python (mainly introduction of `async`/ `await` keyword and inclusion of asyncio into the stdlib), we can do this very easily.

**I have created a gist here with a working example**
https://gist.github.com/shivekkhurana/1de00e1e54c36d250a7f19905fe133b9

It gives  a simple structure to listen to multiple change feeds asynchronously (tested on python 3.6)
Hope it helps.
